### PR TITLE
feat(upgrade): support v2 volume upgrade

### PR DESCRIPTION
**Pull request by derekbit**
#### Which issue(s) this PR fixes:

Issue longhorn/longhorn#6001

### CODE REVIEW

File: pkg/instance/instance.go [hunk 1]
```diff
@@ -702,11 +712,170 @@ func engineResponseToInstanceResponse(e *spdkapi.Engine) *rpc.InstanceResponse {
...
+
+	ops, ok := s.ops[req.DataEngine]
+	if !ok {
+		return nil, grpcstatus.Errorf(grpccodes.Unimplemented, "unsupported data engine %v", req.DataEngine)
+	}
+	return ops.InstanceDeleteTarget(req)
+}
+
+func (ops V1DataEngineInstanceOps) InstanceDeleteTarget(req *rpc.InstanceDeleteTargetRequest) (*emptypb.Empty, error) {
+	return nil, grpcstatus.Error(grpccodes.Unimplemented, "v1 data engine instance target switch over is not supported")
```
Comment (DamiaSan) [inline pkg/instance/instance.go:859]: "v1 data engine instance target delete is not supported"

**Review by DamiaSan — CHANGES_REQUESTED** [review]

File: pkg/instance/instance.go (hunk 1 above)
Comment (derekbit) [reply pkg/instance/instance.go]: Updated, thank you.

**Comment by derekbit** [comment]
@DamiaSan @shuo-wu  Can you help review the PR? Thank you.

cc @innobead

File: pkg/client/instance.go [hunk 2]
```diff
@@ -115,6 +118,17 @@ type InstanceCreateRequest struct {
 	BackendStoreDriver string
 }
 
+type InstanceSuspendRequest struct {
```
Comment (c3y1huang) [inline pkg/client/instance.go:121]: Could you point to where this is used?

**Review by DamiaSan — DISMISSED** [review]

File: pkg/client/instance.go (hunk 2 above)
Comment (derekbit) [reply pkg/client/instance.go]: It is dead code. Removed. Thank you.

**Review by DamiaSan — APPROVED** [review]

File: pkg/instance/instance.go [hunk 3]
```diff
@@ -702,11 +722,170 @@ func engineResponseToInstanceResponse(e *spdkapi.Engine) *rpc.InstanceResponse {
...
+
+	switch req.Type {
+	case types.InstanceTypeEngine:
+		err := c.EngineSwitchOverTarget(req.Name, req.TargetAddress)
+		if err != nil {
+			return nil, grpcstatus.Error(grpccodes.Internal, errors.Wrapf(err, "failed to switch over target for engine %v", req.Name).Error())
+		}
+		return &emptypb.Empty{}, nil
+	case types.InstanceTypeReplica:
+		return nil, grpcstatus.Errorf(grpccodes.InvalidArgument, "target switchover is not supported for instance type %v", req.Type)
```
Comment (c3y1huang) [inline pkg/instance/instance.go:848]:
```suggestion
		return nil, grpcstatus.Errorf(grpccodes.InvalidArgument, "target switch over is not supported for instance type %v", req.Type)
```

File: pkg/instance/instance.go [hunk 4]
```diff
@@ -702,11 +722,170 @@ func engineResponseToInstanceResponse(e *spdkapi.Engine) *rpc.InstanceResponse {
...
+		return nil, grpcstatus.Errorf(grpccodes.InvalidArgument, "unknown instance type %v", req.Type)
+	}
+}
+
+func (s *Server) InstanceDeleteTarget(ctx context.Context, req *rpc.InstanceDeleteTargetRequest) (*emptypb.Empty, error) {
+	logrus.WithFields(logrus.Fields{
+		"name":       req.Name,
+		"type":       req.Type,
+		"dataEngine": req.DataEngine,
+	}).Info("Delete target")
```
Comment (c3y1huang) [inline pkg/instance/instance.go:859]:
```suggestion
	}).Info("Deleting target")
```

**Review by c3y1huang — CHANGES_REQUESTED** [review]

File: pkg/instance/instance.go (hunk 3 above)
Comment (derekbit) [reply pkg/instance/instance.go]: Updated. Thank you.

File: pkg/instance/instance.go (hunk 4 above)
Comment (derekbit) [reply pkg/instance/instance.go]: Updated. Thank you.

**Review by c3y1huang — APPROVED** [review]
Do you think it would be beneficial to separate the (ops V1DataEngineInstanceOps) and (ops V2DataEngineInstanceOps) into different files? Maybe something to consider for future refactoring.

**Comment by derekbit** [comment]
> Do you think it would be beneficial to separate the (ops V1DataEngineInstanceOps) and (ops V2DataEngineInstanceOps) into different files? Maybe something to consider for future refactoring.

Yes, same thought. I will improve it after v1.7.0.

**Pull request merged**